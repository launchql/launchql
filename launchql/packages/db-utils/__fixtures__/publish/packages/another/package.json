{
  "name": "another",
  "version": "0.0.1",
  "description": "skitch project",
  "author": "Dan Lynch <pyramation@gmail.com>"
}
