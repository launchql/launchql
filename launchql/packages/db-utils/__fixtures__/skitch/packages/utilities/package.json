{
  "name": "pg-utilities",
  "version": "0.0.1",
  "description": "PostgreSQL utilities",
  "author": "Dan Lynch <pyramation@gmail.com>",
  "private": true,
  "scripts": {
    "test": "FAST_TEST=1 launchql-templatedb && jest",
    "test:watch": "FAST_TEST=1 launchql-templatedb && jest --watch"
  },
  "devDependencies": {
    "@types/jest": "21.1.0",
    "@types/node": "8.0.0",
    "babel-cli": "6.24.1",
    "babel-jest": "20.0.3",
    "babel-preset-react-app": "3.0.0",
    "dotenv": "5.0.1",
    "jest": "20.0.4"
  },
  "dependencies": {
    "pg": "6.4.0",
    "pg-promise": "6.10.3",
    "@launchql/db-testing": "latest",
    "uuid": "3.1.0"
  }
}