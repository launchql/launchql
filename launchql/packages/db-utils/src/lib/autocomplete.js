import { filter } from 'fuzzy';

export const makeAutocompleteFunctionWithInput = (keys) => (answers, input) => {
  input = input || '';
  return new Promise(function (resolve) {
    setTimeout(function () {
      var fuzzyResult = filter(input, keys);
      resolve(
        fuzzyResult.map(function (el) {
          return el.original;
        })
      );
    }, 25);
  });
};
