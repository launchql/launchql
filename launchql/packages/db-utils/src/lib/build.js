import { listModules, getExtensionsAndModulesChanges } from './modules';
import { dirname, basename, resolve } from 'path';
import { skitchPath as sPath } from './paths';
import { readFileSync } from 'fs';

export const build = async (project) => {
  const skitchPath = await sPath();
  const modules = await listModules();
  const modulesWithChanges = await getExtensionsAndModulesChanges(project);
  const native = [];

  const extensions = Object.keys(modules).reduce((m, key) => {
    const mod = modules[key];
    m[key] = {
      ...mod,
      sql: readFileSync(
        resolve(`${skitchPath}/${mod.path}/sql/${key}--${mod.version}.sql`)
      )
        .toString()
        .split('\n')
        .filter((l, i) => i !== 0)
        .join('\n')
    };
    return m;
  }, {});

  const deps = Object.keys(extensions).reduce((m, k) => {
    m[k] = extensions[k].requires;
    return m;
  }, {});

  // https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
  function dep_resolve(sqlmodule, resolved, unresolved) {
    unresolved.push(sqlmodule);
    let edges = deps[sqlmodule];
    if (!edges) {
      native.push(sqlmodule);
      edges = deps[sqlmodule] = [];
    }
    for (let i = 0; i < edges.length; i++) {
      const dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${sqlmodule}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }
    resolved.push(sqlmodule);
    const index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index);
  }

  const resolved = [];
  const unresolved = [];

  dep_resolve(project, resolved, unresolved);

  const sql = [];

  resolved.forEach((extension) => {
    if (native.includes(extension)) {
      sql.push(`CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`);
    } else {
      sql.push(extensions[extension].sql);
    }
  });
  return sql.join('\n');
};
