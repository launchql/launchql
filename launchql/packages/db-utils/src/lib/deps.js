import { readFileSync } from 'fs';
import { relative } from 'path';
import { sync as glob } from 'glob';
import { listModules } from './modules';
import { getExtensionName } from './extensions';

const makeKey = (sqlmodule) => '/deploy/' + sqlmodule + '.sql';

export const getDeps = async (packageDir) => {
  const extname = await getExtensionName(packageDir);

  var external = [];

  var deps = {};

  // https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
  function dep_resolve(sqlmodule, resolved, unresolved) {
    unresolved.push(sqlmodule);
    let edges = deps[makeKey(sqlmodule)];
    if (!edges) {
      if (/:/.test(sqlmodule)) {
        external.push(sqlmodule);
        edges = deps[sqlmodule] = [];
      } else {
        throw new Error(`no module ${sqlmodule}`);
      }
    }

    for (var i = 0; i < edges.length; i++) {
      var dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${sqlmodule}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }

    resolved.push(sqlmodule);
    var index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index);
  }

  var files = await glob(`${packageDir}/deploy/**/**.sql`);

  for (var i = 0; i < files.length; i++) {
    const data = readFileSync(files[i]);
    var lines = data.toString().split('\n');
    const key = '/' + relative(packageDir, files[i]);
    deps[key] = [];

    for (var j = 0; j < lines.length; j++) {
      var m = lines[j].match(/^-- requires: (.*)/);
      if (m) {
        deps[key].push(m[1].trim());
      }

      // check only:
      var m2;
      var keyToTest;

      if (/:/.test(lines[j])) {
        m2 = lines[j].match(/^-- Deploy ([^:]*):([\w\/]+) to pg/);
        if (m2) {
          keyToTest = m2[2];
          if (extname !== m2[1]) {
            throw new Error(
              'referencing bad project name inside of deploy file\n' + lines[j]
            );
          }
          if (key != makeKey(keyToTest)) {
            throw new Error(
              'deployment script in wrong place or is named wrong internally\n' +
                lines[j]
            );
          }
        }
      } else {
        m2 = lines[j].match(/^-- Deploy (.*) to pg/);
        if (m2) {
          keyToTest = m2[1];
          if (key != makeKey(keyToTest)) {
            throw new Error(
              'deployment script in wrong place or is named wrong internally\n' +
                lines[j]
            );
          }
        }
      }
    }
  }

  let resolved = [];
  var unresolved = [];

  deps = Object.assign(
    {
      [makeKey('apps/index')]: Object.keys(deps)
        .filter((dep) => dep.match(/^\/deploy\//))
        .map((dep) => dep.replace(/^\/deploy\//, '').replace(/.sql$/, ''))
    },
    deps
  );

  dep_resolve('apps/index', resolved, unresolved);
  var index = resolved.indexOf('apps/index');
  resolved.splice(index);
  delete deps[makeKey('apps/index')];

  // move extensions up
  // TODO maybe deprecate this?
  const extensions = resolved.filter((a) => a.match(/^extensions/));
  const normalSql = resolved.filter((a) => !a.match(/^extensions/));
  // console.log({ extensions, normalSql });
  // resolved = useExtensions ? [...extensions, ...normalSql] : [...normalSql];
  resolved = [...extensions, ...normalSql];
  // resolved = [...normalSql];

  return {
    external,
    resolved,
    deps
  };
};

export const extDeps = async (name) => {
  const modules = await listModules();
  const external = [];
  if (!modules[name]) {
    throw new Error(`module ${name} does not exist!`);
  }
  // build deps
  const deps = Object.keys(modules).reduce((memo, key) => {
    memo[key] = modules[key].requires;
    return memo;
  }, {});

  // https://www.electricmonk.nl/log/2008/08/07/dependency-resolving-algorithm/
  function dep_resolve(sqlmodule, resolved, unresolved) {
    unresolved.push(sqlmodule);
    let edges = deps[sqlmodule];
    if (!edges) {
      external.push(sqlmodule);
      edges = deps[sqlmodule] = [];
    }

    for (var i = 0; i < edges.length; i++) {
      var dep = edges[i];
      if (!resolved.includes(dep)) {
        if (unresolved.includes(dep)) {
          throw new Error(`Circular reference detected ${sqlmodule}, ${dep}`);
        }
        dep_resolve(dep, resolved, unresolved);
      }
    }

    resolved.push(sqlmodule);
    var index = unresolved.indexOf(sqlmodule);
    unresolved.splice(index);
  }

  const resolved = [];
  var unresolved = [];

  dep_resolve(name, resolved, unresolved);

  return { external, resolved };
};
