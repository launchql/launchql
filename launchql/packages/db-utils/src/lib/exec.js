import { exec as shellExec } from 'shelljs';
import { PGUSER, PGPASSWORD, PGHOST, PGPORT } from '@launchql/db-env';

export const execSync = (cmd, opts) => {
  shellExec(cmd, {
    env: {
      ...process.env,
      PGUSER,
      PGPASSWORD,
      PGHOST,
      PGPORT
    },
    ...opts
  });
};
