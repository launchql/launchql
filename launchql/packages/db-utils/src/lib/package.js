import { sqitchPath as path } from './paths';
import { getExtensionName } from './extensions';
import { parse, deparse } from 'pgsql-parser';
import { resolve, resolveWithPlan } from './resolve';
import { sync as mkdirp } from 'mkdirp';
import { relative } from 'path';
import { transformProps } from '@launchql/db-transform';
import { writeFileSync, readFileSync } from 'fs';
import { sync as rimraf } from 'rimraf';

const noop = () => undefined;

export const cleanTree = (tree) => {
  return transformProps(tree, {
    stmt_len: noop,
    stmt_location: noop,
    location: noop
  });
};

// default to using the plan, so we stay in sync with the actual order that was developed in code
export const packageModule = async ({
  usePlan = true,
  extension = true
} = {}) => {
  const sqitchPath = await path();
  const resolveFn = usePlan ? resolveWithPlan : resolve;
  const sql = await resolveFn(sqitchPath);
  const extname = await getExtensionName(sqitchPath);

  // sql
  try {
    const parsed = parse(sql);
    const stmts = parsed.reduce((m, node) => {
      if (extension) {
        if (node.RawStmt.stmt.hasOwnProperty('TransactionStmt')) return m;
        if (node.RawStmt.stmt.hasOwnProperty('CreateExtensionStmt')) return m;
      }

      return [...m, node];
    }, []);
    const topLine = extension
      ? `\\echo Use "CREATE EXTENSION ${extname}" to load this file. \\quit\n`
      : '';
    const finalSql = deparse(stmts);
    const tree1 = stmts;
    const tree2 = parse(finalSql);
    const results = {
      sql: `${topLine}${finalSql}`
    };
    const diff =
      JSON.stringify(cleanTree(tree1)) !== JSON.stringify(cleanTree(tree2));
    if (diff) {
      results.diff = true;
      results.tree1 = JSON.stringify(cleanTree(tree1), null, 2);
      results.tree2 = JSON.stringify(cleanTree(tree2), null, 2);
    }

    return results;
  } catch (e) {
    console.error(e);
  }
};

export const writePackage = async ({
  version,
  extension = true,
  usePlan = true,
  sqitchPath
}) => {
  if (!sqitchPath) {
    sqitchPath = await path();
  }
  const pkgPath = `${sqitchPath}/package.json`;
  const pkg = require(pkgPath);
  const extname = await getExtensionName(sqitchPath);
  const makePath = `${sqitchPath}/Makefile`;
  const controlPath = `${sqitchPath}/${extname}.control`;
  const sqlFileName = `${extname}--${version}.sql`;

  const Makefile = readFileSync(makePath).toString();
  const control = readFileSync(controlPath).toString();

  const { sql, diff, tree1, tree2 } = await packageModule({
    extension,
    usePlan
  });

  const outPath = extension ? `${sqitchPath}/sql` : `${sqitchPath}/out`;

  rimraf(outPath);
  mkdirp(outPath);

  if (extension) {
    // control file
    writeFileSync(
      controlPath,
      control.replace(
        /default_version = '[0-9\.]+'/,
        `default_version = '${version}'`
      )
    );

    // package json
    pkg.version = version;
    writeFileSync(pkgPath, JSON.stringify(pkg, null, 2));

    // makefile
    var regex = new RegExp(extname + '--[0-9.]+.sql');
    writeFileSync(makePath, Makefile.replace(regex, sqlFileName));
  }

  if (diff) {
    console.error(
      `DIFF exists! Careful. Check ${relative(sqitchPath, outPath)}/ folder...`
    );
    // writeFileSync(`${outPath}/orig.${sqlFileName}.tree.json`, tree1);
    // writeFileSync(`${outPath}/parsed.${sqlFileName}.tree.json`, tree2);
  }

  const writePath = `${outPath}/${sqlFileName}`;
  writeFileSync(writePath, sql);
  console.log(`${relative(sqitchPath, writePath)} written`);
};
