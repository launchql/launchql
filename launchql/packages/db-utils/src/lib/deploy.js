import * as shell from 'shelljs';
import { resolve } from 'path';

import { listModules } from './modules';

import { skitchPath } from './paths';
import { extDeps } from './deps';

import { PGUSER, PGPASSWORD, PGHOST, PGPORT, PATH } from '@launchql/db-env';

const pg = require('pg');

// should we be parsing the plan file?
// currently assuming only extensions in control file...

export const deploy = async (name, database, opts) => {
  const modules = await listModules();
  const path = await skitchPath();
  if (!modules[name]) {
    throw new Error(`module ${name} does not exist!`);
  }
  const extensions = await extDeps(name);

  const pgPool = new pg.Pool({
    connectionString: `postgres://${PGUSER}:${PGPASSWORD}@${PGHOST}:${PGPORT}/${database}`
  });

  for (let i = 0; i < extensions.resolved.length; i++) {
    const extension = extensions.resolved[i];
    try {
      if (extensions.external.includes(extension)) {
        console.log(`CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`);
        await pgPool.query(
          `CREATE EXTENSION IF NOT EXISTS "${extension}" CASCADE;`
        );
      } else {
        console.log(modules[extension].path);
        console.log(`sqitch deploy db:pg:${database}`);
        const cmd = shell.exec(`sqitch deploy db:pg:${database}`, {
          cwd: resolve(path, modules[extension].path),
          env: {
            PGUSER,
            PGPASSWORD,
            PGHOST,
            PGPORT,
            PATH
          }
        });
        if (cmd.code !== 0) {
          throw new Error('deploy failed');
        }
      }
    } catch (e) {
      pgPool.end();
      process.exit(1);
    }
  }
  pgPool.end();

  return extensions;
};
