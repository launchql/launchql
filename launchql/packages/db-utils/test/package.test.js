import { packageModule } from '../src/lib/package';
process.env.SKITCH_PATH = __dirname + '/../__fixtures__/skitch';
process.env.SQITCH_PATH =
  __dirname + '/../__fixtures__/skitch/packages/secrets';

const clean = (t) =>
  t
    .split('\n')
    .map((a) => a.trim())
    .filter((a) => a)
    .join('\n');

it('creates an extension', async () => {
  const { sql } = await packageModule();
  expect(clean(sql)).toMatchSnapshot();
});
it('creates an extension via plan', async () => {
  const { sql } = await packageModule({ usePlan: true });
  expect(clean(sql)).toMatchSnapshot();
});
