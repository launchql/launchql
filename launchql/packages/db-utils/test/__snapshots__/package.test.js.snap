// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates an extension 1`] = `
"\\\\echo Use \\"CREATE EXTENSION secrets\\" to load this file. \\\\quit
CREATE FUNCTION secretfunction (  ) RETURNS text AS $EOFCODE$
select * from generate_secret();
$EOFCODE$ LANGUAGE sql STABLE;"
`;

exports[`creates an extension via plan 1`] = `
"\\\\echo Use \\"CREATE EXTENSION secrets\\" to load this file. \\\\quit
CREATE FUNCTION secretfunction (  ) RETURNS text AS $EOFCODE$
select * from generate_secret();
$EOFCODE$ LANGUAGE sql STABLE;"
`;
